src/hotspot/share/gc/shenandoah/shenandoahCardStats.cpp

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2022, Amazon.com, Inc. or its affiliates.  All rights reserved.
+ *
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+
+#include "precompiled.hpp"
+
+#include "gc/shenandoah/shenandoahCardStats.hpp"
+#include "logging/log.hpp"
+
+#ifndef PRODUCT
+void ShenandoahCardStats::update_run_work(bool record) {
+  assert(!(_last_dirty || _last_clean) || (_last_dirty && _dirty_run > 0) || (_last_clean && _clean_run > 0),
+         "dirty/clean run stats inconsistent");
+  assert(_dirty_run == 0 || _clean_run == 0, "Both shouldn't be non-zero");
+  if (_dirty_run > _max_dirty_run) {
+    assert(_last_dirty, "Error");
+    _max_dirty_run = _dirty_run;
+  } else if (_clean_run > _max_clean_run) {
+    assert(_last_clean, "Error");
+    _max_clean_run = _clean_run;
+  }
+  _dirty_card_cnt += _dirty_run;
+  _clean_card_cnt += _clean_run;
+
+  // Update local stats
+  {
+    assert(_dirty_run <= _cards_in_cluster, "Error");
+    assert(_clean_run <= _cards_in_cluster, "Error");
+    // Update global stats for distribution of dirty/clean run lengths
+    _local_card_stats[DIRTY_RUN].add((double)_dirty_run*100/(double)_cards_in_cluster);
+    _local_card_stats[CLEAN_RUN].add((double)_clean_run*100/(double)_cards_in_cluster);
+
+    if (record) {
+      // Update global stats for distribution of dirty/clean card %ge
+      _local_card_stats[DIRTY_CARDS].add((double)_dirty_card_cnt*100/(double)_cards_in_cluster);
```

Comment:
typo? `%` -> `a`

Reply:
I mean percentage where I said `%ge`. I'll clarify the comments a bit more. Please continue the review and I'll improve some of the documentation comments for clarity.

Reply:
Got it - I had `age` on my brain.

Reply:
Slightly reworded some of the code comments for greater clarity.